File: common/off_hours_test.go
package common

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestIsOffHoursBlockedAt(t *testing.T) {
	t.Parallel()

	// Use a fixed location for consistent testing
	loc := time.FixedZone("Test", 0)

	tests := []struct {
		name          string
		cfg           OffHoursConfig
		checkTime     time.Time
		wantBlocked   bool
		wantMessage   string
		wantUnblockAt *time.Time
	}{
		{
			name:        "empty config - not blocked",
			cfg:         OffHoursConfig{},
			checkTime:   time.Date(2024, 1, 15, 10, 0, 0, 0, loc), // Monday 10:00
			wantBlocked: false,
		},
		{
			name: "same-day window - inside window",
			cfg: OffHoursConfig{
				Message: "Go to sleep!",
				Windows: []OffHoursWindow{
					{Start: "09:00", End: "17:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 12, 0, 0, 0, loc), // Monday 12:00
			wantBlocked:   true,
			wantMessage:   "Go to sleep!",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 17, 0, 0, 0, loc)),
		},
		{
			name: "same-day window - before window",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "09:00", End: "17:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 15, 8, 0, 0, 0, loc), // Monday 08:00
			wantBlocked: false,
		},
		{
			name: "same-day window - after window",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "09:00", End: "17:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 15, 18, 0, 0, 0, loc), // Monday 18:00
			wantBlocked: false,
		},
		{
			name: "same-day window - at start boundary",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "09:00", End: "17:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 9, 0, 0, 0, loc), // Monday 09:00
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 17, 0, 0, 0, loc)),
		},
		{
			name: "same-day window - at end boundary",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "09:00", End: "17:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 15, 17, 0, 0, 0, loc), // Monday 17:00
			wantBlocked: false,
		},
		{
			name: "overnight window - after start (same day)",
			cfg: OffHoursConfig{
				Message: "Sleep time!",
				Windows: []OffHoursWindow{
					{Start: "23:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 23, 30, 0, 0, loc), // Monday 23:30
			wantBlocked:   true,
			wantMessage:   "Sleep time!",
			wantUnblockAt: ptr(time.Date(2024, 1, 16, 7, 0, 0, 0, loc)),
		},
		{
			name: "overnight window - before end (next day)",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "23:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 16, 5, 0, 0, 0, loc), // Tuesday 05:00
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 16, 7, 0, 0, 0, loc)),
		},
		{
			name: "overnight window - outside window (afternoon)",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "23:00", End: "07:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 15, 14, 0, 0, 0, loc), // Monday 14:00
			wantBlocked: false,
		},
		{
			name: "overnight window - at end boundary",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "23:00", End: "07:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 16, 7, 0, 0, 0, loc), // Tuesday 07:00
			wantBlocked: false,
		},
		{
			name: "day-specific window - matching day",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Days: []string{"monday", "wednesday"}, Start: "03:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 4, 0, 0, 0, loc), // Monday 04:00
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 7, 0, 0, 0, loc)),
		},
		{
			name: "day-specific window - non-matching day",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Days: []string{"monday", "wednesday"}, Start: "03:00", End: "07:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 16, 4, 0, 0, 0, loc), // Tuesday 04:00
			wantBlocked: false,
		},
		{
			name: "day-specific overnight window - morning belongs to previous day",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Days: []string{"tuesday"}, Start: "23:00", End: "07:00"},
				},
			},
			// Tuesday 05:00 should NOT be blocked because the morning portion
			// belongs to Monday's window (which isn't in Days)
			checkTime:   time.Date(2024, 1, 16, 5, 0, 0, 0, loc), // Tuesday 05:00
			wantBlocked: false,
		},
		{
			name: "day-specific overnight window - evening portion blocked",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Days: []string{"tuesday"}, Start: "23:00", End: "07:00"},
				},
			},
			// Tuesday 23:30 should be blocked (evening portion of Tuesday's window)
			checkTime:     time.Date(2024, 1, 16, 23, 30, 0, 0, loc), // Tuesday 23:30
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 17, 7, 0, 0, 0, loc)), // Wednesday 07:00
		},
		{
			name: "day-specific overnight window - next morning blocked",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Days: []string{"tuesday"}, Start: "23:00", End: "07:00"},
				},
			},
			// Wednesday 05:00 should be blocked (morning portion of Tuesday's window)
			checkTime:     time.Date(2024, 1, 17, 5, 0, 0, 0, loc), // Wednesday 05:00
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 17, 7, 0, 0, 0, loc)), // Wednesday 07:00
		},
		{
			name: "day-specific window - case insensitive",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Days: []string{"Monday", "WEDNESDAY"}, Start: "03:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 4, 0, 0, 0, loc), // Monday 04:00
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 7, 0, 0, 0, loc)),
		},
		{
			name: "multiple windows - first matches",
			cfg: OffHoursConfig{
				Message: "Take a break",
				Windows: []OffHoursWindow{
					{Start: "12:00", End: "13:00"},
					{Start: "03:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 12, 30, 0, 0, loc), // Monday 12:30
			wantBlocked:   true,
			wantMessage:   "Take a break",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 13, 0, 0, 0, loc)),
		},
		{
			name: "multiple windows - second matches",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "12:00", End: "13:00"},
					{Start: "03:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 5, 0, 0, 0, loc), // Monday 05:00
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 7, 0, 0, 0, loc)),
		},
		{
			name: "multiple windows - none match",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "12:00", End: "13:00"},
					{Start: "03:00", End: "07:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 15, 10, 0, 0, 0, loc), // Monday 10:00
			wantBlocked: false,
		},
		{
			name: "invalid time format - not blocked",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "invalid", End: "07:00"},
				},
			},
			checkTime:   time.Date(2024, 1, 15, 5, 0, 0, 0, loc),
			wantBlocked: false,
		},
		{
			name: "default message when not specified",
			cfg: OffHoursConfig{
				Windows: []OffHoursWindow{
					{Start: "03:00", End: "07:00"},
				},
			},
			checkTime:     time.Date(2024, 1, 15, 4, 0, 0, 0, loc),
			wantBlocked:   true,
			wantMessage:   "Time to rest!",
			wantUnblockAt: ptr(time.Date(2024, 1, 15, 7, 0, 0, 0, loc)),
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			status := IsOffHoursBlockedAt(tt.checkTime, tt.cfg)

			assert.Equal(t, tt.wantBlocked, status.Blocked, "blocked mismatch")

			if tt.wantBlocked {
				assert.Equal(t, tt.wantMessage, status.Message, "message mismatch")
				require.NotNil(t, status.UnblockAt, "expected unblock time")
				assert.Equal(t, *tt.wantUnblockAt, *status.UnblockAt, "unblock time mismatch")
			} else {
				assert.Nil(t, status.UnblockAt, "expected no unblock time")
			}
		})
	}
}

func TestParseTimeOfDay(t *testing.T) {
	t.Parallel()

	tests := []struct {
		input      string
		wantHour   int
		wantMinute int
		wantErr    bool
	}{
		{"00:00", 0, 0, false},
		{"09:30", 9, 30, false},
		{"23:59", 23, 59, false},
		{"12:00", 12, 0, false},
		{"invalid", 0, 0, true},
		{"25:00", 0, 0, true},
		{"12:60", 0, 0, true},
		{"", 0, 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			t.Parallel()

			hour, minute, err := parseTimeOfDay(tt.input)
			if tt.wantErr {
				assert.Error(t, err)
			} else {
				assert.NoError(t, err)
				assert.Equal(t, tt.wantHour, hour)
				assert.Equal(t, tt.wantMinute, minute)
			}
		})
	}
}

func ptr[T any](v T) *T {
	return &v
}

File: frontend/src/lib/__tests__/offHours.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { evaluateOffHours, isTimeInWindow, type OffHoursConfig, type OffHoursWindow } from '../offHours'

describe('isTimeInWindow', () => {
  it('should return not blocked for invalid time format', () => {
    const window: OffHoursWindow = { start: 'invalid', end: '07:00' }
    const t = new Date('2024-01-15T04:00:00')
    const result = isTimeInWindow(t, window)
    expect(result.blocked).toBe(false)
    expect(result.unblockAt).toBe(null)
  })

  it('should block during same-day window', () => {
    const window: OffHoursWindow = { start: '09:00', end: '17:00' }
    const t = new Date('2024-01-15T12:00:00') // Monday at noon
    const result = isTimeInWindow(t, window)
    expect(result.blocked).toBe(true)
    expect(result.unblockAt).not.toBe(null)
    expect(result.unblockAt?.getHours()).toBe(17)
    expect(result.unblockAt?.getMinutes()).toBe(0)
  })

  it('should not block outside same-day window', () => {
    const window: OffHoursWindow = { start: '09:00', end: '17:00' }
    const t = new Date('2024-01-15T08:00:00') // Monday at 8am
    const result = isTimeInWindow(t, window)
    expect(result.blocked).toBe(false)
  })

  it('should block during overnight window (before midnight portion)', () => {
    const window: OffHoursWindow = { start: '23:00', end: '07:00' }
    const t = new Date('2024-01-15T23:30:00') // Monday at 11:30pm
    const result = isTimeInWindow(t, window)
    expect(result.blocked).toBe(true)
    expect(result.unblockAt).not.toBe(null)
    expect(result.unblockAt?.getDate()).toBe(16) // Next day
    expect(result.unblockAt?.getHours()).toBe(7)
  })

  it('should block during overnight window (after midnight portion)', () => {
    const window: OffHoursWindow = { start: '23:00', end: '07:00' }
    const t = new Date('2024-01-16T03:00:00') // Tuesday at 3am
    const result = isTimeInWindow(t, window)
    expect(result.blocked).toBe(true)
    expect(result.unblockAt).not.toBe(null)
    expect(result.unblockAt?.getDate()).toBe(16) // Same day
    expect(result.unblockAt?.getHours()).toBe(7)
  })

  it('should not block outside overnight window', () => {
    const window: OffHoursWindow = { start: '23:00', end: '07:00' }
    const t = new Date('2024-01-15T12:00:00') // Monday at noon
    const result = isTimeInWindow(t, window)
    expect(result.blocked).toBe(false)
  })

  it('should respect day-specific windows', () => {
    const window: OffHoursWindow = { days: ['monday'], start: '09:00', end: '17:00' }
    
    // Monday - should block
    const monday = new Date('2024-01-15T12:00:00')
    expect(isTimeInWindow(monday, window).blocked).toBe(true)
    
    // Tuesday - should not block
    const tuesday = new Date('2024-01-16T12:00:00')
    expect(isTimeInWindow(tuesday, window).blocked).toBe(false)
  })

  it('should handle overnight window with day-specific config', () => {
    // Window starts on Monday night, ends Tuesday morning
    const window: OffHoursWindow = { days: ['monday'], start: '23:00', end: '07:00' }
    
    // Monday at 11:30pm - should block (start day matches)
    const mondayNight = new Date('2024-01-15T23:30:00')
    expect(isTimeInWindow(mondayNight, window).blocked).toBe(true)
    
    // Tuesday at 3am - should block (previous day was Monday)
    const tuesdayMorning = new Date('2024-01-16T03:00:00')
    expect(isTimeInWindow(tuesdayMorning, window).blocked).toBe(true)
    
    // Tuesday at 11:30pm - should not block (Tuesday not in days list)
    const tuesdayNight = new Date('2024-01-16T23:30:00')
    expect(isTimeInWindow(tuesdayNight, window).blocked).toBe(false)
    
    // Wednesday at 3am - should not block (previous day was Tuesday)
    const wednesdayMorning = new Date('2024-01-17T03:00:00')
    expect(isTimeInWindow(wednesdayMorning, window).blocked).toBe(false)
  })

  it('should handle empty days array as all days', () => {
    const window: OffHoursWindow = { days: [], start: '09:00', end: '17:00' }
    const t = new Date('2024-01-15T12:00:00')
    expect(isTimeInWindow(t, window).blocked).toBe(true)
  })

  it('should handle undefined days as all days', () => {
    const window: OffHoursWindow = { start: '09:00', end: '17:00' }
    const t = new Date('2024-01-15T12:00:00')
    expect(isTimeInWindow(t, window).blocked).toBe(true)
  })
})

describe('evaluateOffHours', () => {
  it('should return not blocked when not enabled', () => {
    const config: OffHoursConfig = { enabled: false, blocked: false }
    const result = evaluateOffHours(new Date(), config)
    expect(result.blocked).toBe(false)
    expect(result.message).toBe('')
  })

  it('should return not blocked when no windows configured', () => {
    const config: OffHoursConfig = { enabled: true, blocked: false, windows: [] }
    const result = evaluateOffHours(new Date(), config)
    expect(result.blocked).toBe(false)
  })

  it('should return blocked with default message when in window', () => {
    const config: OffHoursConfig = {
      enabled: true,
      blocked: false,
      windows: [{ start: '00:00', end: '23:59' }],
    }
    const result = evaluateOffHours(new Date('2024-01-15T12:00:00'), config)
    expect(result.blocked).toBe(true)
    expect(result.message).toBe('Time to rest!')
  })

  it('should return blocked with custom message when in window', () => {
    const config: OffHoursConfig = {
      enabled: true,
      blocked: false,
      message: 'Go to sleep!',
      windows: [{ start: '00:00', end: '23:59' }],
    }
    const result = evaluateOffHours(new Date('2024-01-15T12:00:00'), config)
    expect(result.blocked).toBe(true)
    expect(result.message).toBe('Go to sleep!')
  })

  it('should find earliest unblock time across multiple windows', () => {
    const config: OffHoursConfig = {
      enabled: true,
      blocked: false,
      windows: [
        { start: '03:00', end: '08:00' },
        { start: '03:00', end: '06:00' },
      ],
    }
    const t = new Date('2024-01-15T04:00:00')
    const result = evaluateOffHours(t, config)
    expect(result.blocked).toBe(true)
    expect(result.unblockAt?.getHours()).toBe(6) // Earlier of the two
  })

  it('should not block when outside all windows', () => {
    const config: OffHoursConfig = {
      enabled: true,
      blocked: false,
      windows: [
        { start: '03:00', end: '07:00' },
        { start: '22:00', end: '23:00' },
      ],
    }
    const t = new Date('2024-01-15T12:00:00')
    const result = evaluateOffHours(t, config)
    expect(result.blocked).toBe(false)
  })
})